src/read_data.c

```diff
@@ -34,21 +34,30 @@
 #if HAVE_ZLIB
 #include <zlib.h>
 #endif
+#include <float.h>
 #include <stddef.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 #include <stdio.h>
 #include <math.h>
 #include <time.h>
 
-#define READ_DATA_NOSWAP(T)                                           \
+#define READ_DATA_NOSWAP(T, TT)                                       \
     do {                                                              \
+        TT min_ = (TT)READ_TYPE_MIN;                                  \
+        TT max_ = (TT)READ_TYPE_MAX;                                  \
         const size_t block_size = READ_BLOCK_SIZE / data_size;        \
         if ( len <= block_size ) {                                    \
             readcount = fread(v, data_size, len, (FILE *)mat->fp);    \
             if ( readcount == len ) {                                 \
                 for ( i = 0; i < len; i++ ) {                         \
-                    data[i] = (T)v[i];                                \
+                    TT val_ = v[i];                                   \
+                    if (val_ >= min_ && val_ <= max_) {               \
+                        data[i] = (T)val_;                            \
+                    } else {                                          \
+                        break;                                        \
+                    }                                                 \
```

Comment:
So I've just realized a problem... when this cast on line 57 is doing float -> int, let's say `uint8_t` specifically, the valid range of floats that can be legally cast to `uint8_t` are [-0.5, 255.5] (I forget if inclusive or exclusive), but `READ_TYPE_MIN` is 0 and `READ_TYPE_MAX` is 255, so we would in fact start rejecting some valid cases.

Not sure best way to do this...

Reply:
For integer conversion you can read http://tzimmermann.org/2018/04/20/safe-integer-conversion-in-c/.
For floating-point to integer conversion we need to deal with rounding. There is https://en.cppreference.com/w/c/numeric/math/round in C99, which rounds to nearest integer.

Reply:
@tbeu want me to just take my best stab at it?

Reply:
Sure. That would be appreciated.

Reply:
I gave it a try, but have just realised what I've committed is rather wrong... will continue tomorrow.